crates/orderbook/src/fee_subsidy.rs

```diff
@@ -0,0 +1,214 @@
+pub mod config;
+pub mod cow_token;
+
+use crate::order_quoting::QuoteParameters;
+use anyhow::Result;
+use ethcontract::U256;
+use futures::future;
+use std::sync::Arc;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Subsidy {
+    /// A flat discount nominated in the native token to discount from fees.
+    ///
+    /// Flat fee discounts are applied **before** the fee factor.
+    pub discount: f64,
+
+    /// Minimum fee amount after applying the flat subsidy. This prevents flat
+    /// fee discounts putting the fee amount below 0.
+    ///
+    /// Flat fee discounts are applied **before** the fee factor.
+    pub min_discounted: f64,
+
+    /// An additional fee factor.
+    pub factor: f64,
+}
+
+impl Default for Subsidy {
+    fn default() -> Self {
+        Self {
+            discount: 0.,
+            min_discounted: 0.,
+            factor: 1.,
+        }
+    }
+}
+
+/// Trait for describing fee subsidies applied to orders.
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait FeeSubsidizing: Send + Sync {
+    async fn subsidy(&self, parameters: QuoteParameters) -> Result<Subsidy>;
+}
+
+/// Combine multiple fee subsidy strategies into one!
+pub struct FeeSubsidies(pub Vec<Arc<dyn FeeSubsidizing>>);
+
+#[async_trait::async_trait]
+impl FeeSubsidizing for FeeSubsidies {
+    async fn subsidy(&self, parameters: QuoteParameters) -> Result<Subsidy> {
+        let subsidies = future::try_join_all(
+            self.0
+                .iter()
+                .map(|strategy| strategy.subsidy(parameters.clone())),
+        )
+        .await?;
+
+        Ok(subsidies
+            .into_iter()
+            .fold(Subsidy::default(), |a, b| Subsidy {
+                discount: a.discount + b.discount,
+                min_discounted: a.min_discounted.max(b.min_discounted),
+                factor: a.factor * b.factor,
+            }))
```

Comment:
This is important and a little confusing. I think the motivation for this change is to generalize over the different kind of subsidies we had before so now we have an implementation of this trait for FeeSubsidyConfiguration, CowSubsidy and FeeSubsidies.
I find this confusing because it is not obvious for example that min_discounted would be folded through max instead of min. And that we would combine the subsidies in this way first versus individually applying them in order. To me it is more natural to have one component that does the final subsidy without generalizing subcomponents through traits. CowSubsidy only wants to apply a factor but because of the trait it has to return a whole Subsidy struct.
I don't think we are going to have many varied FeeSubsidizing implementations.

Not very strong opinion so I don't mind merging this anyway but this was my first gut feeling. Might be helpful to have a comment somewhere explaining how these types fit together.

Reply:
> I think the motivation for this change is to generalize over the different kind of subsidies we had before so now we have an implementation of this trait for FeeSubsidyConfiguration, CowSubsidy and FeeSubsidies.

Exactly. It was mostly to have `CowSubsidy` and `FeeSubsidy` be modelled by a single component.

> And that we would combine the subsidies in this way first versus individually applying them in order.

Yes, I agree. Initially I had something like:
```rust
trait FeeSubsidizing {
    fn apply_subsidy(fee_in_eth: f64) -> f64;
}
```

As it was the more natural interface IMO. What made me change, is that by using a `Subsidy` struct and having `FeeSubsidizing` combining into a `Subsidy` struct allows us to simultaneously fetch the subsidy while waiting for price, gas, native price estimates to complete. Otherwise, we would have had to first do all the estimates and only then compute the subsidy. I'm OK with that model as well, but wanted to keep the amount of additional waiting for quotes as low as possible.

> Not very strong opinion so I don't mind merging this anyway but this was my first gut feeling. Might be helpful to have a comment somewhere explaining how these types fit together.

Sounds good! I'll add a module-level explanation about why we do things this way.

Reply:
Added module-level comment explaining design and clarification why we use `max` for the `min_discounted` parameter.